README.md
# ChromeHeadless

<p align="left">
<img src="https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-trigger/badge/icon" alt="building">

<img src="https://img.shields.io/readthedocs/pip.svg" alt="doc on">
</p>
A lightweight built-in for website profiling.


**I am still ACTIVELY building and testing it, this is just my side crawling project for a specific task**

- [x] web-version-based crawling -pup_run.js
- [x] mobile-based-crawling -mobile.js
- [x] distributed-crawling using shared memory counter -task_dispatch.c
- [ ] add JS tracking
- [ ] documentation

## Install

Install chrome and check whether it is stable:

```
#install chrome debian
sudo apt-get update
sudo apt-get install google-chrome-stable

### get your chrome version
apt list --installed | grep chrome

#my version is that google-chrome-stable/now 62.0.3202.89-1 amd64
\ [installed,upgradable to: 64.0.3282.186-1]

```

Directly run headless chrome to get screenshot and dump html page, in separate steps

```
//take screen
google-chrome --headless --disable-gpu —-timeout=300 --enable-logging --v=1  \
--screenshot --window-size=1280,1696  http://facebook.com

//PS: --enable-logging --v=1 will log all the intermediate data, \
// you have to parse the log date to get the final URL.

//dump source
google-chrome --headless --disable-gpu --dump-dom https://www.chromestatus.com/


google-chrome --headless --hide-scrollbars --remote-debugging-port=9222 --disable-gpu &


google-chrome --headless  --disable-gpu --window-size=1280,1696
--screenshot https://www.google.com

```

Or install npm dependence to have more functionalities:
```
uname -a
Linux XXXX 3.13.0-132-generic #181-Ubuntu SMP Wed Sep 13 13:25:03 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux


google-chrome --version
Google Chrome 62.0.3202.89


npm -v
5.6.0

nodejs -v
v8.5.0

npm install puppeteer
// google provided library
```

## Get screenshots, HTML source and redirections.

One command line for all, paralleled version is under construction :star:

The url is parsed into the parameter.

```
nodejs pup_run.js --url=http://facebook.com
```

## Test and Samples:

### 1) Single case test

Time for get all on facebook page: 1.7s.

It captures the redirection from http to https.

```
time nodejs pup_run.js --url=http://facebook.com
[1] Redirection: from url:<http://facebook.com> to <https://www.facebook.com/>
[1] Redirection Chain was successfully saved.
[2] HTML was successfully saved.
[3] Screen was saved.

node pup_run.js --url=http://facebook.com  0.96s user 0.18s system 64% cpu 1.756 total
```

- The screenshot: <img src="https://github.com/ririhedou/ChromeHeadless/blob/master/fb/facebook.com.screen.png" width="800" height="400" />


- The redirection chain:

```
<http://facebook.com> -> <https://www.facebook.com/>
```

- The html source: [TEXT](https://github.com/ririhedou/ChromeHeadless/blob/master/fb/facebook.com.source.txt)

### 2) A large scale test

```
//--dir output directory
time node pup_run.js --file=_home_datashare_dns_history_20170906_facebook.com.out --dir=fbtest/

For 225 test urls, the total time consumption is 335 seconds (1.5s/per url).

For 3165 test urls, the total time consumption is 763.21s user 130.13s system 17% cpu 1:27:09.00 total (1.65s/per url).

For 11483 test urls, the total time consumption 2968.55s user 516.68s system 19% cpu 4:55:55.49 (1.54s/per url).

```


### 3) Distributed version :rocket::rocket::rocket::rocket::rocket::rocket:

Use the task_dispatch to distribute the crawling tasks.

```
gcc task_dispatcher.c --std=c99
./a.out start_idx stop_idx

nohup ./a.out 1 10 &

#run subdomains
nohup node subdomain_run.js --dir=fb/ > subdomain.out &

pstree:
a_cyclops.out─┬─a_cyclops.out───sh───node─┬─chrome─┬─chrome─┬─2*[chrome───14*[{chrome}]]
              │                           │        │        ├─chrome───20*[{chrome}]
              │                           │        │        ├─2*[chrome───21*[{chrome}]]
              │                           │        │        ├─chrome───24*[{chrome}]
              │                           │        │        └─chrome───44*[{chrome}]
              │                           │        ├─chrome───6*[{chrome}]
              │                           │        └─24*[{chrome}]
              │                           └─9*[{node}]
              ├─a_cyclops.out───sh───node─┬─chrome─┬─chrome─┬─3*[chrome───14*[{chrome}]]
              │                           │        │        ├─chrome───27*[{chrome}]
              │                           │        │        ├─chrome───22*[{chrome}]
              │                           │        │        └─chrome───29*[{chrome}]
              │                           │        ├─chrome───150*[{chrome}]
              │                           │        └─22*[{chrome}]
              │                           └─9*[{node}]
              ├─a_cyclops.out───sh───node─┬─chrome─┬─chrome─┬─chrome───14*[{chrome}]
              │                           │        │        └─chrome───16*[{chrome}]
              │                           │        ├─chrome───6*[{chrome}]
              │                           │        └─22*[{chrome}]
              │                           └─9*[{node}]
              └─a_cyclops.out───sh───node─┬─chrome─┬─chrome─┬─4*[chrome───14*[{chrome}]]
                                          │        │        ├─2*[chrome───20*[{chrome}]]
                                          │        │        └─chrome───23*[{chrome}]
                                          │        ├─chrome───6*[{chrome}]
                                          │        └─22*[{chrome}]
                                          └─9*[{node}]

```

## Tutorials (I read)

Usage:

- use headless chrome to profile website. [LINK](https://medium.com/@e_mad_ehsan/getting-started-with-puppeteer-and-chrome-headless-for-web-scrapping-6bf5979dee3e)
- scrapy with Puppeteer [LINK](https://codeburst.io/a-guide-to-automating-scraping-the-web-with-javascript-chrome-puppeteer-node-js-b18efb9e9921)
- Puppeteer API reference [LINK](https://github.com/GoogleChrome/puppeteer/blob/v0.10.2/docs/api.md#pageclose)

How to build a :shipit:distributed:shipit: system:

- ChromeLess [LINK](https://github.com/graphcool/chromeless)
- Brozzler [LINK](https://github.com/internetarchive/brozzler)


## FAQ

- What if there are too many renders? (pkill chrome or run 100 urls each time)
- Error in CentOS 7? (https://github.com/GoogleChrome/puppeteer/issues/290)
  try --no-sandbox
- How to kill zombie processes [LINK](https://www.geeksforgeeks.org/zombie-processes-prevention/)

- Cannot stop for url http://www.makereadyweb.com/,  (pkill -9 chrome every 3 hours)


More code is on the way


Changelog
2018-04-10: version 0.15
========================
add todo list


2018-03-26: version 0.15
========================
Pass testing on my local machine on both domains and subdomains
transform to the server


2018-03-25: version 0.14
========================
add mobile setting
using iPhone 6

mobile.js
//author ketian
//ririhedou@gmail.com

module.exports =
{
    run_mobile_multiple_urls : run_mobile_multiple_urls
}


const fs = require('fs');
const puppeteer = require('puppeteer');
const devices = require('puppeteer/DeviceDescriptors');
const globalDir = './'

// let us emulate Iphone6:
async function run_mobile_single(myUrl){

    const browser = await puppeteer.launch({ignoreHTTPSErrors: true, args: ['--no-sandbox']});
    //if you do not use this, error will happen on CentOS 7

    const iPhone = devices['iPhone 6'];

    var page = null;

        page = await browser.newPage();

        await page.emulate(iPhone);

        //await page.setUserAgent(mobile_userAgent);

        var arrayOfStrings = String(myUrl).split('/');

        let name = globalDir + arrayOfStrings[arrayOfStrings.length-1];

        console.log("The store location is at " + name);

        await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

        const url = await page.url();

        console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

        var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

        await fs.writeFile(name + '.redirect', directionChain, (err) => {
            if (err) throw err;
            console.log('[1] Redirection Chain was successfully saved.');
        });

        let HTML = await page.content(); //content is enough for the HTML content

        var filepath = name + '.source.txt';
        await fs.writeFile(filepath, HTML, (err) => {
            if (err) throw err;
            console.log('[2] HTML was successfully saved.');
        });

        //await page.setViewport({width: 1600, height: 800});
        await page.screenshot({path: name + '.screen.png'});

        console.log('[3] Screen was saved.');
        await page.close();


    await browser.close();
};

async function run_mobile_multiple_urls(urlList, beginId, globalDir){

  var page = null;

  const browser = await puppeteer.launch({args: ['--no-sandbox']});
  //if you do not use this, error will happen on CentOS 7

  const iPhone = devices['iPhone 6'];

  var arrayLength = urlList.length;

  for (var i = 0; i < arrayLength; i++) {

    //the i-th element
    var myUrl = urlList[i];

    console.log("[TEST MOBILE]Visiting  "+ beginId + ":" + myUrl);

    page = await browser.newPage();

    await page.emulate(iPhone);

    try {
            var arrayOfStrings = String(myUrl).split('/');

            let name = globalDir + arrayOfStrings[arrayOfStrings.length-1];

            await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

            //await page.waitForNavigation();

            const url = await page.url();

            //console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

            var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

            await fs.writeFile(name + '.redirect', directionChain, (err) => {
                if (err) throw err;
                //console.log('[1] Redirection Chain was successfully saved.');
            });

            let HTML = await page.content(); //content is enough for the HTML content

            var filepath = name + '.source.txt';
            await fs.writeFile(filepath, HTML, (err) => {
                if (err) throw err;
                //console.log('[2] HTML was successfully saved.');
            });

            await page.screenshot({path: name + '.screen.png'});

            //console.log('[3] Screen was saved.');
            console.log("[SUCC MOBILE]Done on "+  beginId + ":" + myUrl);
        }
    catch (e)
        {
            //console.log(e);
            console.log('[FAIL MOBILE]Cannot go to for ' +  beginId + ":" + myUrl + ', continue...');
        }

    await page.close();

  }
  await browser.close();

};


//var url ="http://facebook.com";
//run_mobile_single(url);

data_process.js
//author ketian
//ririhedou@gmail.com

module.exports =
{
    readSquatting : readSquatting
}

//Define a struture that contains the domain, squat_from, type
const Struct = (...keys) => ((...v) => keys.reduce((o, k, i) => {o[k] = v[i]; return o} , {}))
const Item = Struct('suspicious_domain', 'squat_from', 'type', 'url')

const fs = require('fs');

function readSquatting(filename)
{
    var lines = fs.readFileSync(filename, 'utf-8').split('\n').filter(Boolean);
    var domain_structure_arr = new Array();

    for (var i in lines)
        {
          //console.log(lines[i].length);
          var arr = lines[i].split(",");

          //do some data cleaning
          arr[0] = arr[0].replace('(u\'','').replace('\'','');
          arr[1] = arr[1].replace('\'','').replace('\'','').replace(' ','');
          var url = 'http://' + arr[1];

          arr[3] = arr[3].replace('u\'','').replace('\'','').replace('\'','').replace(' ','');

          domain_structure_arr.push(Item(String(arr[1]), arr[0], arr[3], url));
        }
    /*
    for (var j in domain_structure_arr)
    {
       console.log(domain_structure_arr[j]);
    }*/

    return domain_structure_arr;
};


function readSubdomainSquatting()
{

}
//var filepath = '/home/ketian/Desktop/squatting_domains/domain_collect/_home_datashare_dns_history_20170906_4chan.org.out'
//filepath ='/home/ketian/Desktop/squatting_domains/domain_collect/_home_datashare_dns_history_20170906_bloomberg.com.out'
//readSqautting(filepath);

subdomain_run.js
//author ketian
//ririhedou@gmail.com

const fs = require('fs');
const puppeteer = require('puppeteer');
const devices = require('puppeteer/DeviceDescriptors');

var argv = require('minimist')(process.argv.slice(2));


function checkDirectorySync(directory) {
  try {

    fs.statSync(directory);
    console.log("Existing "+ directory + ", no need to recreate");

  } catch(e) {

    fs.mkdirSync(directory);
    console.log("Create a directory as "+ directory);
  }
}

//terminate the process in case some weird things happen
process.on('unhandledRejection', up => { throw up })

/*
*  run multiple URL instances
*  @input: list of URLs
*
*/
async function run_multiple_urls(urlList, _idx, globalDir){

  var page = null;

  const browser = await puppeteer.launch({args: ['--no-sandbox']});
  //if you do not use this, error will happen on CentOS 7

  var arrayLength = urlList.length;

  for (var i = 0; i < arrayLength; i++) {

    //the i-th element
    myUrl = urlList[i];

    console.log("[TEST]Visiting  "+ _idx + ":" + myUrl);

    page = await browser.newPage();
    try {
            var arrayOfStrings = String(myUrl).split('/');

            var name = globalDir + arrayOfStrings[arrayOfStrings.length-1];

            await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

            //await page.waitForNavigation();

            const url = await page.url();

            //console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

            var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

            await fs.writeFile(name + '.redirect', directionChain, (err) => {
                if (err) throw err;
                //console.log('[1] Redirection Chain was successfully saved.');
            });

            let HTML = await page.content(); //content is enough for the HTML content

            var filepath = name + '.source.txt';
            await fs.writeFile(filepath, HTML, (err) => {
                if (err) throw err;
                //console.log('[2] HTML was successfully saved.');
            });

            await page.setViewport({width: 1600, height: 800});
            await page.screenshot({path: name + '.screen.png'});

            //console.log('[3] Screen was saved.');
            console.log("[SUCC]Done on "+  _idx + ":" + myUrl);
        }
    catch (e)
        {
            //console.log(e);
            console.log('[FAIL]Cannot go to for ' +  _idx + ":" + myUrl + ', continue...');
        }
    await page.close();

  }
  await browser.close();

};

/*
*  run multiple URL instances in mobile version
*  @input: list of URLs
*
*/

async function run_mobile_multiple_urls(urlList, _idx, globalDir){

  var page = null;

  const browser = await puppeteer.launch({args: ['--no-sandbox']});
  //if you do not use this, error will happen on CentOS 7

  const iPhone = devices['iPhone 6'];

  var arrayLength = urlList.length;

  for (var i = 0; i < arrayLength; i++) {

    //the i-th element
    var myUrl = urlList[i];

    console.log("[TEST MOBILE]Visiting  "+ _idx + ":" + myUrl);

    page = await browser.newPage();

    await page.emulate(iPhone);

    try {
            var arrayOfStrings = String(myUrl).split('/');

            let name = globalDir + arrayOfStrings[arrayOfStrings.length-1];

            await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

            //await page.waitForNavigation();

            const url = await page.url();

            //console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

            var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

            await fs.writeFile(name + '.redirect', directionChain, (err) => {
                if (err) throw err;
                //console.log('[1] Redirection Chain was successfully saved.');
            });

            let HTML = await page.content(); //content is enough for the HTML content

            var filepath = name + '.source.txt';
            await fs.writeFile(filepath, HTML, (err) => {
                if (err) throw err;
                //console.log('[2] HTML was successfully saved.');
            });

            await page.screenshot({path: name + '.screen.png'});

            //console.log('[3] Screen was saved.');
            console.log("[SUCC MOBILE]Done on "+  _idx + ":" + myUrl);
        }
    catch (e)
        {
            //console.log(e);
            console.log('[FAIL MOBILE]Cannot go to for ' +  _idx + ":" + myUrl + ', continue...');
           }

    await page.close();

  }
  await browser.close();

};


async function run_file_in_subdomain_fashion(filename, _idx, outputDir, output_mobile)
{

    var lines = fs.readFileSync(filename, 'utf-8').split('\n').filter(Boolean);
    var urlList = new Array();

    for (var i in lines)
        {
          var arr = lines[i].split(",");
          //do some data cleaning
          arr[3] = arr[3].replace('(u\'','').replace(')','').replace('\'','').replace('\'','').replace(' ','');
          var url = 'http://' + arr[3];
          urlList.push(url);

        }

    console.log('[STAT]Total length of ' + filename + '  is   ' + String(urlList.length));

    var length = urlList.length;
    var interval = 10;
    for (var i = 0; i < length; i = i+interval)
    {
        var subArray = urlList.slice(i,i+interval);
        try
        {
            await run_multiple_urls(subArray, _idx, outputDir);
            await run_mobile_multiple_urls(subArray, _idx, output_mobile);
        }
        catch (e)
        {
           console.log("[FXXK]Browser launch failed");
        }
    }

}


//Read all files into a map
async function main() {

    const sub = "subdomain";
    var outputDir = null;

    if (sub.length > 0)
    {  outputDir = argv.dir + sub + '/'|| './' }
    else
    {  outputDir = argv.dir || './' }

    console.log("globalDir is " + outputDir);
    fs.mkdirSync(outputDir);
    var files = fs.readdirSync('./subdomain_collect/');
    files.sort();
    for (i in files)
    {
        var file = './subdomain_collect/' + files[i];
        var _id = i;
        var output = outputDir + _id.toString() + '/';
        var output_mobile = outputDir + _id.toString() + '_mobile/';
        checkDirectorySync(output);
        checkDirectorySync(output_mobile);
        console.log(_id, file);
        await run_file_in_subdomain_fashion(file, _id.toString(), output, output_mobile);
    }

    process.exit(0);
}

main();

//run_file_in_subdomain_fashion(_map);
//node subdomain_run.js

task_dispatcher.c
//author ketian
//ririhedou@gmail.com

#define _XOPEN_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/ipc.h>
#include <sys/shm.h>

#define WATHEVER_CMD_PREFIX "node pup_run.js --dir=fb/ --id=%d"

#define PROC_MAX 5
//The maximum processes the machine can tolerate
//10X Fast

//you can customize the work with any task.
void work(int i)
{
    char buf[1024];
    snprintf(buf, 1024, WATHEVER_CMD_PREFIX, i);
    //use snprintf to avoid buffer overflow
    system(buf);
}

//sample command as:
//./a.out 1 4
int main(int argc, char** argv)
{
    //int ppid = getpid();
    if (argc<3)
    {
        //usage();
        exit(-1);
    }
    int start_val = atoi(argv[1]);
    int stop_val = atoi(argv[2]);

    int shmid = shmget(IPC_PRIVATE, sizeof(int), IPC_CREAT | 0666);
    if (shmid<0)
    {
        exit(-1);
    }

    signal(SIGCHLD, SIG_IGN);
    //to prevent zombie processes

    int* counter;
    counter = shmat(shmid, NULL, 0);
    *counter = 0;

    for (int i=start_val; i<=stop_val; i++)
    {
        int cpid;
        //retry counter
        int rcnt = 0;
        //respwaning new task, increase counter by one
        while(!__sync_bool_compare_and_swap(counter,*counter, (*counter) +1));
retry:
        cpid = fork();
        rcnt++;
        if (cpid==0)
        {
            work(i);
            while(!__sync_bool_compare_and_swap(counter,*counter, (*counter)-1));
            printf("[SPONGEBOB]Done %d counter=%d\n", i, *counter);
            shmdt(counter);
            exit(-1);
        }
        else if (cpid==-1)
        {
            //failed to fork...
            printf("[SPONGEBOB]Fail to fork on id=%d...\n", i);
            if (rcnt>100)
            {
                goto out; //we re-try 100 times
            }
            goto retry;
        }
out:
        while (*counter>=PROC_MAX)
        {
            sleep(10);
        }
    }

    printf("[SPONGEBOB] Done and exit");
    exit(0);
}


pup_run.js
//author ketian
//ririhedou@gmail.com

const fs = require('fs');
const puppeteer = require('puppeteer');
var argv = require('minimist')(process.argv.slice(2));

const beginUrl = argv.url;
const beginFile = argv.file;
const beginId = (argv.id.toString() || '').toString();

var outputDir = null;
var mobileDir = null;

if (beginId.length > 0)
{
    outputDir = argv.dir + beginId + '/'|| './';
    mobileDir = argv.dir + beginId + '_mobile/' || './';
}
else
{
    outputDir = argv.dir || './';
    mobileDir = argv.dir + '_mobile`' || './';
}

const globalDir = outputDir;
const globalMobileDir = mobileDir;

console.log("globalDir is " + globalDir);
console.log("globalMobile Dir is" + globalMobileDir);

function checkDirectorySync(directory) {
  try {

    fs.statSync(directory);
    console.log("Existing " + directory + ", no need to recreate");
    process.exit();

  } catch(e) {
    fs.mkdirSync(directory);
    console.log("Create a directory as " + directory);
  }
}

checkDirectorySync(globalDir);
checkDirectorySync(globalMobileDir);

//terminate the process in case some weird things happen
process.on('unhandledRejection', up => { throw up })

/*
* run for only 1 url at each time
*/
async function run_single(myUrl){

    const browser = await puppeteer.launch({args: ['--no-sandbox']});
    //if you do not use this, error will happen on CentOS 7
    var page = null;
    try
    {
        page = await browser.newPage();

        var arrayOfStrings = String(myUrl).split('/');

        let name =  arrayOfStrings[arrayOfStrings.length-1];

        console.log("The store location is at " + name);

        await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

        const url = await page.url();

        console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

        var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

        await fs.writeFile(name + '.redirect', directionChain, (err) => {
            if (err) throw err;
            console.log('[1] Redirection Chain was successfully saved.');
        });

        let HTML = await page.content(); //content is enough for the HTML content

        var filepath = name + '.source.txt';
        await fs.writeFile(filepath, HTML, (err) => {
            if (err) throw err;
            console.log('[2] HTML was successfully saved.');
        });

        await page.setViewport({width: 1600, height: 800});
        await page.screenshot({path: name + '.screen.png'});

        console.log('[3] Screen was saved.');
        await page.close();
    }
    catch(e)
        {
          console.log('Cannot go to for ' + myUrl + ', stop the browser...');
        }

    await browser.close();
};

/*
*  run multiple URL instances
*  @input: list of URLs
*
*/
async function run_multiple_urls(urlList){

  var page = null;

  const browser = await puppeteer.launch({args: ['--no-sandbox']});
  //if you do not use this, error will happen on CentOS 7

  var arrayLength = urlList.length;

  for (var i = 0; i < arrayLength; i++) {

    //the i-th element
    myUrl = urlList[i];

    console.log("[TEST]Visiting  "+ beginId + ":" + myUrl);

    page = await browser.newPage();
    try {
            var arrayOfStrings = String(myUrl).split('/');

            let name = globalDir + arrayOfStrings[arrayOfStrings.length-1];

            await page.goto(myUrl,  { timeout: 5000 }); //5000 is 5000ms

            //await page.waitForNavigation();

            const url = await page.url();

            //console.log('[1] Redirection: from url:<' + myUrl + '> to <' + url + '>');

            var directionChain = 'Redirection: from url:<' + myUrl + '> to <' + url + '>';

            await fs.writeFile(name + '.redirect', directionChain, (err) => {
                if (err) throw err;
                //console.log('[1] Redirection Chain was successfully saved.');
            });

            let HTML = await page.content(); //content is enough for the HTML content

            var filepath = name + '.source.txt';
            await fs.writeFile(filepath, HTML, (err) => {
                if (err) throw err;
                //console.log('[2] HTML was successfully saved.');
            });

            await page.setViewport({width: 1600, height: 800});
            await page.screenshot({path: name + '.screen.png'});

            //console.log('[3] Screen was saved.');
            console.log("[SUCC]Done on "+  beginId + ":" + myUrl);
        }
    catch (e)
        {
            //console.log(e);
            console.log('[FAIL]Cannot go to for ' +  beginId + ":" + myUrl + ', continue...');
        }
    await page.close();

  }
  await browser.close();

};


async function run_file_in_separate_fashion(beginFile)
{
    var dp = require('./data_process');

    var mobile = require('./mobile');

    var domain_structure_list = dp.readSquatting(beginFile);
    var urlList = new Array();
    for (i in domain_structure_list)
    {
        //console.log(i, domain_structure_list[i].url);
        //we only monitor non-duplicate cases
        if (urlList.indexOf(domain_structure_list[i].url)==-1)
            urlList.push(domain_structure_list[i].url);
    }

    urlList.sort();
    console.log('[STAT]Total length of ' + beginFile + '  is   ' + String(urlList.length));
    var length = urlList.length;
    var interval = 5;
    for (var i = 0; i < length; i = i+interval)
    {
        var subArray = urlList.slice(i,i+interval)
        try
        {
            await mobile.run_mobile_multiple_urls(subArray, beginId, globalMobileDir);
        }
        catch (e)
        {
           console.log("[FXXK MOBILE]Browser launch failed, lose 10 urls");
        }

        try
        {
            await run_multiple_urls(subArray);
        }
        catch (e)
        {
           console.log("[FXXK]Browser launch failed, lose 10 urls");
        }
    }

    process.exit(0);

}


//The main function to run this
if (beginUrl){
    console.log('Run single url......');
    run_single(beginUrl);
}

else if (beginFile)
{
    console.log('Run single file.......');
    run_file_in_separate_fashion(beginFile);
}

else if (beginId)
{
    var di = require('./map_domain_to_id');
    var filename = './domain_collect/' + di.get_domain_from_id(beginId);
    run_file_in_separate_fashion(filename);
}

else
{
    console.log("Nothing happened");
}
//run_single(beginUrl);
//var urlList = ['http://facebook.com', 'http://google.com', 'http://googggg.vvv'];
//run_multiple_urls(urlList);
//Commands:
//nodejs pup_run.js --file=/home/ketian/Desktop/squatting_domains/domain_collect/_home_datashare_dns_history_20170906_4chan.org.out --dir=test/  | wc -l
//nodejs pup_run.js --file=/home/ketian/Desktop/squatting_domains/domain_collect/_home_datashare_dns_history_20170906_4chan.org.out --dir=test/  | tee 4chan.log